### ISSUE TITLE
Feature request: auto-save instead of manually saving MD files

### ISSUE THREAD

**Issue by anniesexton** [careteditor/caret#870]
I love pretty much everything about Caret – except that I have to manually save each note as a file. I would love to see my notes just autosave, the way they do in Bear/Apple Notes/SimpleNote. Removing this layer of friction makes me so much more likely to use the app.

One of the reasons I abandoned iA writer and other notetaking apps that require you to choose a folder and give each note a title/filename is because at the end of the day, it made it much clunkier quickly take notes on the fly.

Would love to see this feature added!

**Comment by erusev** [comment]
Hi Annie and thanks for getting in touch!

To make sure I understand what you imagine, I'd like to ask you a couple of questions:

1) Where should auto-saved notes be stored?
2) What should be the names of auto-saved notes?

Thanks again!

**Comment by anniesexton** [comment]
Hi @erusev – Thanks for the response!

>Where should auto-saved notes be stored?

I'm not sure how apps like Bear/Ulysses/Apple Notes do it, they might have a database they use, but I imagine you could start by just saving notes in a single directory that sits in whatever cloud service you're using to sync things, iCloud/Dropbox, whatever you like. Folders could be created within that root directory, however the user sees fit.

>What should be the names of auto-saved notes?

Most apps that have this "auto-save" features just use the first line of the note, which, if you're using markdown might often be the first H1 tag. 

I already LOVE how fast Caret is right now, and all the shortcuts it has, but right now the process of saving each note is pretty clunky and makes me not want to use it for jotting down quick ideas. I just want to type something up really quick and move on.

Thanks for considering this! Please let me know if you have any other questions about this idea

**Comment by erusev** [comment]
@anniesexton thanks for elaborating on this!

I thought a bit more on the issue. Creating file is an important part of Caret and you are right that we should make it easier. 

What if we did this:

When you are in an Untitled file and you hit ⌘S, instead of the native MacOS dialog, the app opens a popup, similar to the popup that opens on ⌘P (Go to File). You type a character or two to select the folder you want to save your file to and you hit enter to save it.

The app can use the H1 to generate a name for the file and it can select it in the sidebar on creation so that you can quickly rename it if you don't like the generated name.

How does this sound?

Thanks!

**Comment by anniesexton** [comment]
Hmm – I actually think it's important to entirely remove the step of having to manually save each note, even if it just takes a few key strokes, since it still requires some thought on deciding _where_ to file away each note. For quick note creation, the task of organizing notes into different folders can be done later. If you're familiar with the Getting Things Done productivity systems, the idea of having a main "inbox" to collect thoughts/notes in a single place, unordered, to later process and organize, is really really helpful. I think having a default folder for all notes (probably at the root of where all notes are kept) would be best. Users can easily reorganize notes into different folders if they choose to do so.

Alternatively (or additionally), if user could open a specific folder in Caret and then all notes would automatically be created in that folder. So the folder for each note to be automatically created in would be determined by the folder that's selected in the sidebar (aka, where the user is "standing" at the time).

>The app can use the H1 to generate a name for the file and it can select it in the sidebar on creation so that you can quickly rename it if you don't like the generated name.

^ This part is exactly what I was thinking of, yes! I'd say you could even just rely on the first line, even if its not an H1.

Keep up the great work!! Ya'll are creating a wonderful tool.

**Comment by erusev** [comment]
@anniesexton thanks for elaborating and apologies for taking me so long to respond!

We want to make it easier to create notes and your suggestion makes a lot of sense, but I'm concerned about potential bad side effects on other use cases, like the use case for temporary documents.

We feel like a lot of people use Untitled documents to write temporary texts like a responses an emails or a blog posts. These are temporary texts because once people write them, they copy them to their email client / blogging platform / etc. and then they discard them.

With the current design, discarding an Untitled document is easy and intuitive - you just hit ⌘W (and then enter to confirm).  With a design that auto-creates files, the same would be more difficult. You'll have to select the file in the sidebar and then delete it and if, for some reason, you close the tab that points to the file, discarding that file would be even more difficult because now you'll have to browse your files until you find the files and then delete it.

I'm looking for a design that makes creating files easier and doesn't make discarding temporary files more difficult.

What if the ⌘S popup (which would look similar to the ⌘P popup), defaulted to the current folder? Then you can do ⌘S and Enter and you'll have your file in the current folder with the name based on H1 or the first line. ⌘S and Enter - that's it.

Do you feel this would still be too much friction?

Thanks!

**Comment by anniesexton** [comment]
@erusev I know we're kinda getting into the nitty gritty of personal use-cases here, but this is something I've thought a lot about, and I see your point. I love talking about this stuff :)

Obviously I am just one data point here, but in my opinion, I don't think that easy deletion and auto-saving are mutually exclusive. Take a look at how Apple Notes does it: you can easily create a note w/o having to fumble through making a decision on where that note should live. In fact, I'm not even sure if notes from Apple Notes even exist as individual files, they might just be stored in a database somewhere and then rendered by the app. Not quite sure. But the point is, you can create a note with a single keyboard shortcut/click AND you can do the same for deleting it. To delete a note, you can literally just click on the note and hit DELETE. They provide you with the ability to sort notes into folders as well, a very simply drag and drop menu on the left hand side, very handy.

In short: I agree that there shouldn't be hardly any friction in deleting notes, but I'm not a fan of using temporary documents as the solution. For one thing, the app doesn't know if a file should be temporary as not, so I'd say its always safer to save the file. 

>You'll have to select the file in the sidebar and then delete it and if, for some reason, you close the tab that points to the file, discarding that file would be even more difficult because now you'll have to browse your files until you find the files and then delete it.

^ With the current UI for showing the file system in Caret, this is a valid concern. This could be solved in a couple ways:

* Providing a very easy universal search functionality – just a keyboard shortcut that allows them to search all notes
* Displaying all folders that contain Caret notes (see Apple Notes) so users don't have to poke around all over their hard drive to find where they put the note.

I've been viewing the potential for Caret is as a self-contained note taking app, where files don't mingle around your computer with other files. Or at least, that's how I would _like_ to use it in the future. Think Evernote, Apple Notes, Bear, etc. In these apps, the "file system" you're working with is entirely displayed in the interface – I very much prefer this. You're not actually saving files to your hard drive the way you typically would with other documents.

The way Caret operates now is more like a document editor, like a typical markdown editor, or even like a word processor like Word or Google Docs. I think a huge reason note-taking apps evolved the way they have was because there was a need for a quicker way to create/delete/manage files, which is what lead them away from relying on the typical flow of creating a documents in a filesystem.

Does that make sense? So while I would love to use Caret is one way – more like a self-contained note-taking app that allows for quick creation/deletion of notes – you're the product owners, and if you'd like to see it more as a document editor, that is completely your prerogative. 

Happy to keep discussing this stuff, I'm kind of productivity nerd and love talking about UX design about note-taking apps 🤣

**Comment by erusev** [comment]
> Happy to keep discussing this stuff, I'm kind of productivity nerd and love talking about UX design about note-taking apps 🤣

Same here :)

Can we take this to email? I'm at: hello [at] erusev [dot] com.

**Comment by sguergachi** [comment]
I'd like to pop in and say I completely agree with @anniesexton. If you consider the flow of writing short form, eliminating the friction of saving has a huge impact to maintaining momentum, the last thing you want to think about is where to save the thing, all you care about is the thing you wrote being saved *somewhere* and easily retrievable later. I also *hate* seeing all my files named "Untitled", it's like we are stuck in the old Windows 98 days where computer software couldn't be smart enough to give my file a more contextually relevant title based on what I wrote other than "Untitled", such a wasted opportunity to delight users... (at *minimum* put a timestamp in there so I can contextually discern what a text file might be about)

Long form writing works better in the mode Caret, Word, etc. but that's not how I use Caret and not how I imagine most people user Caret. I imagine Caret is where writing/ideas starts an evolves but probably ends up in Word or something more industrial strength for book length work for example.

I also want to add that I like Caret for the fact it works with plain text files in the back-end and not a database, because text files are portable and infinitely malleable.

**Comment by erusev** [comment]
I agree. We need a much better workflow for saving and better names for Untitled files. We have some ideas in mind. Hopefully, we'll be able to implement them for our [upcoming product release](https://twitter.com/careteditor/status/1136198029357264896).
